execution/executor/src/components/in_memory_state_calculator.rs

```diff
@@ -41,62 +41,65 @@ impl IntoLedgerView for TreeState {
-            let state_cache = checkpoint_state_view.into_state_cache();
-            let calculator = InMemoryStateCalculator::new(
-                &checkpoint_state,
-                state_cache,
+
+            let state = if self.num_transactions == checkpoint_next_version {
+                checkpoint_state
+            } else {
+                ensure!(
+                self.num_transactions - checkpoint_next_version <= MAX_WRITE_SETS_AFTER_CHECKPOINT,
+                "Too many versions after state checkpoint. checkpoint_next_version: {}, num_transactions: {}",
                 checkpoint_next_version,
+                self.num_transactions,
             );
-            calculator.calculate_for_write_sets_after_checkpoint(&write_sets)?
-        };
+                let checkpoint_state_view = CachedStateView::new(
+                    StateViewId::Miscellaneous,
+                    db.clone(),
+                    self.num_transactions,
+                    checkpoint_state.checkpoint.clone(),
+                    Arc::new(SyncProofFetcher::new(db.clone())),
+                )?;
+                let write_sets =
+                    db.get_write_sets(checkpoint_next_version, self.num_transactions)?;
+                checkpoint_state_view.prime_cache_by_write_set(&write_sets)?;
+                let state_cache = checkpoint_state_view.into_state_cache();
+                let calculator = InMemoryStateCalculator::new(
+                    &checkpoint_state,
+                    state_cache,
+                    checkpoint_next_version,
+                );
+                calculator.calculate_for_write_sets_after_checkpoint(&write_sets)?
+            };
 
-        let transaction_accumulator = Arc::new(InMemoryAccumulator::new(
-            self.ledger_frozen_subtree_hashes,
-            self.num_transactions,
-        )?);
+            let transaction_accumulator = Arc::new(InMemoryAccumulator::new(
+                self.ledger_frozen_subtree_hashes,
+                self.num_transactions,
+            )?);
 
-        Ok(ExecutedTrees::new(state, transaction_accumulator))
+            Ok(ExecutedTrees::new(state, transaction_accumulator))
+        } else {
+            Err(anyhow!(
+                    "state_checkpoint_next_version is None due to overflow. Current state_checkpoint_version: {:?}",
+                    self.state_checkpoint_version,
+                ))
+        }
     }
 }
 
 const MAX_WRITE_SETS_AFTER_CHECKPOINT: LeafCount = 200_000;
 
-fn state_checkpoint_next_version(checkpoint_version: Option<Version>) -> Version {
+fn state_checkpoint_next_version(checkpoint_version: Option<Version>) -> Option<Version> {
     match checkpoint_version {
-        None => 0,
-        Some(v) => {
-            if v == PRE_GENESIS_VERSION {
-                0
-            } else {
-                v + 1
-            }
-        }
+        None => Some(0),
+        Some(v) => v.checked_add(1),
```

Comment:
If you insist you'd like to check for overflow, do it in place here, make this function return `Result<Version>`

Reply:
Won't check for overflow given https://github.com/aptos-labs/aptos-core/pull/1619. We will just panic if there is overflow.